debugging-sos-lldb-via-core/test.sh

```diff
@@ -0,0 +1,233 @@
+
+echo "${framework_dir}"
+
+"${framework_dir}"/createdump --name 'coredump.%d' "${exec_pid}" | tee exec.pid
+
+kill "${exec_pid}"
+kill "${run_pid}" || true
+
+coredump="coredump.${exec_pid}"
+test -f "${coredump}"
+
+# Make sure dotnet-sos is not active
+lldb --batch \
+     -c "${coredump}" \
+     --one-line "soshelp" >lldb.out 2>&1
+cat lldb.out
+grep -F "error: 'soshelp' is not a valid command." lldb.out
+
+# Object Inspection
+
+echo "[dumpobj]"
+lldb-core 'dso' > lldb.out
+cat lldb.out
+id=$(grep -F 'System.String[]' lldb.out | head -1 | cut -d' ' -f 2)
+lldb-core "dumpobj ${id}" > lldb.out
+cat lldb.out
+grep 'Array:       Rank 1,' lldb.out
+
+echo "[dso]"
+lldb-core 'dso' > lldb.out
+cat lldb.out
+# TODO: enable this
+# if grep '<unknown type>' lldb.out; then
+#     echo 'fail: <unknown type> found in dso output.'
+#     exit 2
+# fi
+
+echo "[dumpheap]"
+lldb-core "dumpheap -stat" > lldb.out
+cat lldb.out
+# TODO: enable this
+# if grep UNKNOWN lldb.out; then
+#     echo 'fail: UNKNOWN classes found in dumpheap'
+#     exit 2
+# fi
+
+echo "[gcroot]"
+lldb-core 'dso' > lldb.out
+cat lldb.out
+id=$(grep -F 'System.Threading.Tasks' lldb.out | head -1 | cut -d' ' -f 2)
+lldb-core "gcroot ${id}" > lldb.out
+cat lldb.out
+grep 'Found [[:digit:]]* unique roots' lldb.out
+count=$(grep 'Found [[:digit:]]* unique roots' lldb.out | sed -E 's|Found ([[:digit:]]*) unique roots.*|\1|')
+if [[ $count -le 0 ]]; then
+   echo "fail: $count unique roots found"
+   exit 2
+fi
+# TODO: enable
+# if grep -F '<error>' lldb.out; then
+#     echo 'fail: <error> found in gcroot output'
+#     exit 2
+# fi
+
+echo "[pe]"
+
+# Examining code and stacks
+
+echo "[clrthreads]"
+lldb-core "clrthreads" > lldb.out
+cat lldb.out
+# TODO: fail if there are any "Ukn" at all
+# if grep Ukn lldb.out;  then
+#     echo 'fail: Ukn found in clrthreads'
+#     exit 2
+# fi
+
+echo "[ip2md]"
+lldb-core 'clrthreads' > lldb.out
+cat lldb.out
+thread_id=$(grep -A5 'ID OSID ThreadOBJ' lldb.out | tail -4 | grep -v 'Finalizer' | head -1 | awk '{print $1}')
+lldb-core "thread select ${thread_id}" 'clrstack' > lldb.out
+cat lldb.out
+ip=$(grep 'OS Thread Id:' lldb.out -A5 | tail -n3 | grep -v -F '[Prestub' | head -1 | cut -d' ' -f2)
+lldb-core "ip2md ${ip}" > lldb.out
+cat lldb.out
+if ! grep 'IsJitted' lldb.out ; then
+    echo 'sJitted field not found in ip2md'
+    exit 2
+fi
+
+echo "[clru]"
+lldb-core "clru ${ip}" > lldb.out
+cat lldb.out
+if grep 'Unmanaged code' lldb.out; then
+    echo 'fail: clru thinks IP points to unmanaged code'
+    exit 2
+fi
+
+echo "[dumpstack]"
+lldb-core 'clrstack' > lldb.out
+cat lldb.out
+stack_pointer=$(grep 'OS Thread Id' lldb.out -A3 | tail -1 | awk '{ print $1}')
+lldb-core "dumpstack ${stack_pointer}"
+cat lldb.out
+
+echo "[eestack]"
+lldb-core 'eestack' > lldb.out
+cat lldb.out
+grep 'Child-SP         RetAddr          Caller, Callee' lldb.out
```

Comment:
nit: Is the whitespace here going to stay consistent? Should something like \s\+ be used?

Reply:
Thanks! Fixed.